### ISSUE TITLE
append/prepend in file.replace dependent on show_changes=True

### ISSUE THREAD

**Issue by lorengordon** [saltstack/salt#21093]
If `show_changes` is `False`, and `append_if_not_found` or `prepend_if_not_found` are `True`, then `file.replace` will overwrite the file so it only contains the `repl`/`not_found_content` string. This is because the code relies on the `newfile` parameter to write out the changes, but `newfile` is only populated if `show_changes` is `True`.

Behavior is the same on 2014.7, 2015.2, and develop.
https://github.com/saltstack/salt/blob/2014.7/salt/modules/file.py#L1176

```
# echo 'abc' > ~/foo
# salt-call --local file.replace path=~/foo pattern='xyz' repl='xyz' append_if_not_found=True show_changes=False
local:
    True
# cat foo
xyz
```
```
# salt-call --local file.replace path=~/foo pattern='baz' repl='none' not_found_content='some string' append_if_not_found=True show_changes=False
local:
    True
# cat foo
some string
```
I've overhauled much of `file.replace` in PR https://github.com/saltstack/salt/pull/21030, but I haven't yet addressed this behavior. To correct it, I think the append/prepend block might need to read through the file itself.

**Comment by rallytime** [comment]
@lorengordon Thanks very much for all of your fixes and for filing this new report/question. I think your intuition sounds correct, though @basepi or @whiteinge might have some other comments about addressing that behavior.

**Comment by whiteinge** [comment]
@lorengordon +1

**Comment by lorengordon** [comment]
Thought about this some more and realized the simplest fix is to modify the `if` test that gates population of `new_file`:

https://github.com/saltstack/salt/blob/2014.7/salt/modules/file.py#L1163

to read:
```
                if show_changes or append_if_not_found or prepend_if_not_found:
```

**Comment by basepi** [comment]
Did you end up making this change @lorengordon? Can this be closed?

**Comment by lorengordon** [comment]
No, not yet, but this one looks easy and I'll have time tomorrow to knock it out. Would you prefer it fixed in 2014.7 or 2015.2?

**Comment by rallytime** [comment]
@lorengordon Usually we prefer bugs to be fixed on the "earliest supported branch" that this bug occurs on. The earliest supported branch right now is 2014.7, but I haven't investigated this particular bug, so I am not sure where the fix should be made. I see you've made your PR against 2015.2, but we can backport that easily.

**Comment by lorengordon** [comment]
Gotcha. There are now so many changes to file.replace between 2014.7 and 2015.2 that I just wasn't sure that git would handle the merge well. So I figured I'd make the change to 2015.2 first, and that way if you wanted to port it to 2014.7 you could quickly handle any conflict when merging forward just by copying the entire function from 2015.2 (as it would already contain the fix).

**Comment by rallytime** [comment]
@lorengordon That is a very good point, as we are starting to see some bigger merge conflicts as the branches diverge. The fix might be different in 2014.7. Let's leave your pull request as is, and then I can look at applying it to 2014.7 as you've suggested.
